README.md
# x2048

A simple Haskell remake of the famous game 2048. Features an AI for the game.

### Installation

(requires Stack):

    stack build

### Execution

x2048 can be played in two modes: single player and AI.

In single player you experience the classic feel of a 2048 game. You can use the keys `2`,`4`,`6`,`8` to move down, left, right, up. To launch this mode, type in your terminal:

    stack exec x2048-exe

`AI` mode lets an AI play for you! Very cybernetic. However don't expect 4096 tiles to pop nor anything special: at the current stage of development, the AI is _sometimes_ (i.e. 4 times out of 100 games) able to reach the 2048 tile; it is much more likely that the game(s) will end reaching the 1024 or maybe the 512 tile. This is something I'm working on, of course. To launch this mode:

     stack exec x2048-exe ts

You'll be asked how many games you want the AI to play; then, it will proceed with the simulations. For every game, you'll be shown the board reached and the value of the top tile.

ChangeLog.md
# Changelog for x2048

## Unreleased changes


test/Spec.hs
import X2048
import Test.HUnit

main :: IO ()
main = do
    runTestTT tests
    return ()
    


board1 :: Board
board1 = fromString "2222222222222222"

test1 = TestCase (assertEqual "left: " [4,4,0,0,4,4,0,0,4,4,0,0,4,4,0,0] (move board1 "LEFT"))


board2 :: Board
board2 = fromString "0004222402280028"

test2 = TestCase (assertEqual "UP"    [2,4,4,8,0,0,2,16,0,0,0,0,0,0,0,0] (move board2 "UP"))
test3 = TestCase (assertEqual "DOWN"  [0,0,0,0,0,0,0,0,0,0,2,8,2,4,4,16] (move board2 "DOWN"))
test4 = TestCase (assertEqual "LEFT"  [4,0,0,0,4,2,4,0,4,8,0,0,2,8,0,0] (move board2 "LEFT"))

board3 :: Board
board3 = [4,0,0,0,4,0,0,2,8,0,0,0,64,32,0,0]
test5 = TestCase (assertEqual "DOWN" [0,0,0,0,8,0,0,0,8,0,0,0,64,32,0,2] (move board3 "DOWN"))

board4 = [4,0,0,0,0,0,0,0,2,0,0,0,4,0,4,0]
test6 = TestCase (assertEqual "SEI" [0,0,0,0,4,0,0,0,2,0,0,0,4,0,4,0] (move board4 "DOWN"))

board5 = [8,4,32,2,64,32,2,8,32,16,8,0,16,8,2,0]
test7 = TestCase (assertEqual "Sette" [11,15] (freeIndexes board5))

tests = TestList [ TestLabel "test uno: " test1
                 , TestLabel "test due: " test2
                 , TestLabel "test due: " test3
                 , TestLabel "test due: " test4
                 , TestLabel "test due: " test5
                 , TestLabel "test due: " test6
                 , TestLabel "test due: " test7 ]


app/Main.hs
import Data.List
import Data.IORef
import System.IO
import System.Random
import Control.Monad
import System.Environment (getArgs)
import X2048
import AI
 
main = do
    hSetBuffering stdin NoBuffering
    hSetEcho stdin False

    board <- initialBoard
    b <- newIORef board
    
    args <- getArgs
    case args of
        [] -> mainLoop b True
        ["ai"] -> do
                    mainLoopAi b
                    return ()
        ["ts"] -> do
                    hSetEcho stdin True
                    putStrLn "How many tests?"
                    numStr <- getLine
                    let numTests = read numStr :: Int
                    hSetEcho stdin False
                    counter b [] numTests

mainLoop :: IORef Board -> Bool -> IO ()
mainLoop board printOrNot = do
    if printOrNot
        then do
            readIORef board >>= prettyPrint
            putStrLn "================="
        else return ()

    c <- getChar
    
    concreteBoard <- readIORef board
    case c of
        '8' -> writeIORef board (move concreteBoard "UP")
        '6' -> writeIORef board (move concreteBoard "RIGHT")
        '4' -> writeIORef board (move concreteBoard "LEFT")
        '2' -> writeIORef board (move concreteBoard "DOWN")
        _   -> mainLoop board False

    b' <- readIORef board
    
    if (concreteBoard == b')
        then mainLoop board False
        else do
            cpuMove board
            mainLoop board True

mainLoopAi :: IORef Board -> IO Board
mainLoopAi board = do    
    concreteBoard <- readIORef board
    
    let chosenDirection = choice concreteBoard 
    writeIORef board (move concreteBoard chosenDirection)

    b' <- readIORef board

    if   (concreteBoard == b')
    then return b'
    else do
        cpuMove board
        mainLoopAi board

count :: Eq a => [a] -> a -> Int
count [] n = 0
count (x:xs) n = if (x==n) then 1 + count xs n else count xs n
 
counter board maximums iterations = do
    if (length maximums < iterations)
    then do
        endBoard <- mainLoopAi board
        prettyPrint endBoard
        putStrLn $ show $ maximum endBoard
        newConcreteBoard <- initialBoard
        writeIORef board newConcreteBoard
        counter board (maximum endBoard : maximums) iterations
    else do
        putStrLn $ "128: " ++ (show $ count maximums 128)
        putStrLn $ "256: " ++ (show $ count maximums 256)
        putStrLn $ "512: " ++ (show $ count maximums 512)
        putStrLn $ "1024: " ++ (show $ count maximums 1024)
        putStrLn $ "2048: " ++ (show $ count maximums 2048)


src/X2048.hs
module X2048
( Board
, Direction
, prettyPrint
, move
, cpuMove
, freeIndexes
, generateRotations
, replace
, initialBoard
, addNewTile )
where

import Data.IORef
import Data.List (intercalate, transpose)
import Data.List.Split (chunksOf)
import System.Random (randomRIO)

type Board      = [Int]
type Direction  = String

initialBoard :: IO Board
initialBoard = do
    board <- return $ replicate 16 0
    board <- addNewTile board
    board <- addNewTile board
    return board

-- | Replaces element at IDX with newCell.
replace :: Int -> a -> [a] -> [a]
replace idx newCell list = take idx list ++ [newCell] ++ drop (idx+1) list

-- | Merges two cell of a row/column if they are equal whenever possible.
groupEquals :: [Int] -> [Int]
groupEquals [] = []
groupEquals [x] = [x]
groupEquals (xa:xb:xs)
    | xa == xb  = (xa+xa) : groupEquals xs
    | otherwise = xa : groupEquals (xb : xs)

-- | Moves the board to the left, merges cells and then returns the new matrix
mergeLeft :: [Board] -> [Board]
mergeLeft matrix   = map (fillWithZeros . groupEquals) removed
    where removed = map (filter (/=0)) matrix
          fillWithZeros l = l ++ replicate (4 - length l) 0 

-- | Matrix rotations functions
mirror = map reverse
rotl   = transpose . map reverse
rotr   = map reverse . transpose

move :: Board -> Direction -> Board
move board dir
    | dir == "LEFT"  = concat $ mergeLeft matrix
    | dir == "RIGHT" = concat $ mirror $ mergeLeft (mirror matrix)
    | dir == "UP"    = concat $ rotr $ mergeLeft (rotl matrix)
    | dir == "DOWN"  = concat $ rotl $ mergeLeft (rotr matrix)
        where matrix = chunksOf 4 board

generateRotations :: Board -> [ [Board] ]
generateRotations b = [board, mirror board, rotl board, rotr board]
    where board = chunksOf 4 b

-- | Pretty print of Board
prettyPrint :: Board -> IO ()
prettyPrint b = putStrLn $ intercalate "\n" $ map (intercalate "\t") (chunksOf 4 boardStr)
    where boardStr = map show b

-- | Returns the indexes of the empty cells.
freeIndexes :: Board -> [Int]
freeIndexes board = map snd freeTuples
    where indexes = zipWith (\x i->(x,i)) board [0..]
          freeTuples = filter (\(x,i)-> x==0) indexes

-- | Place a new tile and update the IORef containing the board.
cpuMove :: IORef Board -> IO ()
cpuMove iorefBoard = do
    concreteBoard    <- (readIORef iorefBoard)
    newConcreteBoard <- addNewTile concreteBoard
    writeIORef iorefBoard newConcreteBoard

-- | Given a Board, find an empty tile and fill it with 2 or 4.
addNewTile :: Board -> IO Board
addNewTile board = do
    let frees = freeIndexes board

    idx <- randomRIO (0, (length frees)-1) :: IO Int

    newCell <- randomRIO (1, 10) :: IO Int
    newCell <- return $ if (newCell == 1) then 4 else 2

    return $ replace (frees !! idx) newCell board


src/AI.hs
module AI where

import X2048
import Data.Maybe (fromJust)
import Data.List (elemIndex, sort)

-- | Generate every possible CPU moves that could follow the current board.
everyPossibleCpuMove :: Board -> [Board]
everyPossibleCpuMove b = withTwos ++ withFours
    where frees = freeIndexes b
          withTwos  = [b] >>= (\x -> [replace i 2 x | i <- frees])
          withFours = [b] >>= (\x -> [replace i 4 x | i <- frees])

-- | Generates the 4 translation of a board.
--
-- In other words, this function returns the four boards the user can get by moving up, down, left or right.
-- In case the movement in one of such directions is impossible (e.g. no tile can be moved to the left), an empty list will be returned.
boardTranslations :: Board -> [Board]
boardTranslations board = map (\b -> if b == board then [] else b) boards  -- Evita di rimanere fermo
    where boards = map (move board) ["UP", "DOWN", "LEFT", "RIGHT"]

-- | __(ugly)__ Converts from idx to direction. Strictly related to the
-- function boardTranslations.
convertToDir :: Int -> Direction
convertToDir idx = case idx of
                    0 -> "UP"
                    1 -> "DOWN"
                    2 -> "LEFT"
                    3 -> "RIGHT"

-- | Given a board, return the best direction where to move the board according
-- to the AI.
choice :: Board -> Direction
choice board = convertToDir indexOfBest
    where usrChoices    = boardTranslations board :: [Board]
          trees         = map (mkCpuMoveTree 2) usrChoices :: [CpuMoveTree]
          evaluatedDirs = map cpuTreeEval trees :: [Float]
          indexOfBest   = fromJust $ elemIndex (maximum evaluatedDirs) evaluatedDirs

-- | Evaluates a board.
evaluate :: Board -> Float
evaluate board = fromIntegral (emptyBlocks * (monotonicity board)) * (variance board)
    where emptyBlocks = quot (length $ freeIndexes board) 2

-- | Gives a rate of monotonicity.
-- A monotonic row (or column) is made of tiles in order. For example:
-- [2,4,8,16] is a perfectly fine monotonic row;
-- [16,8,4,2] is also monotonic;
-- [4,256,2,0] is not monotonic;
-- [2,2,2,2] is monotonic.
-- The more monotonic a board, the better.
monotonicity :: Board -> Int
monotonicity board = maximum $ map monotony boards
    where boards = generateRotations board
          monotony m = sum $ map (\row -> if (row == (sort row)) then 1 else 0) m 

-- | Calculates the variance of a board.
-- Variances gives the idea of how much tiles of the same row/columns differs from one another. For example:
-- [2,2,2,2] will have the lowest variance
-- [2,4,8,16] will have a greater, but still low variance
-- [1024,1024,1024,1024] will have the lowest variance, again
-- [256,2,4,2] will have a big variance
-- The smaller a variance value, the better.
variance :: Board -> Float
variance board = (fromIntegral 4) / (fromIntegral (1 + var))
    where avg = sum board `quot` (length board)
          var = sum $ map (\c -> abs (avg - c)) board


data UserMoveTree = UserMoveTree Board [CpuMoveTree]
                  | UserMoveLeaf Board deriving (Show)
                  
data CpuMoveTree  = CpuMoveTree Board [UserMoveTree]
                  | CpuMoveLeaf Board deriving (Show)

mkCpuMoveTree :: Int -> Board -> CpuMoveTree
mkCpuMoveTree 0 b = CpuMoveLeaf b
mkCpuMoveTree n b = CpuMoveTree b $ map (mkUserMoveTree (n-1)) (boardTranslations b)

mkUserMoveTree :: Int -> Board -> UserMoveTree
mkUserMoveTree 0 b = UserMoveLeaf b
mkUserMoveTree n b = UserMoveTree b $ map (mkCpuMoveTree (n-1)) (everyPossibleCpuMove b)

userTreeEval :: UserMoveTree -> Float
userTreeEval (UserMoveLeaf b)   = evaluate b 
userTreeEval (UserMoveTree b l) = evaluate b + ((sum $ map cpuTreeEval l) / (fromIntegral $ (1 + length l)))
-- userTreeEval (UserMoveTree b l) = evaluate b + (median $ map cpuTreeEval l)

cpuTreeEval :: CpuMoveTree -> Float
cpuTreeEval (CpuMoveLeaf b)   = evaluate b 
cpuTreeEval (CpuMoveTree b l) = evaluate b + ((sum $ map userTreeEval l) / (fromIntegral $ (1 + length l)))
-- cpuTreeEval (CpuMoveTree b l) = evaluate b + (median $ map userTreeEval l)

